DOC-12351 Add or Remove Services and Rebalance (#3846)
couchbase/docs-server: modules/rest-api/pages/rest-set-up-services-existing-nodes.adoc

= Assigning Services on an Existing Node

:description: pass:q[Use the REST API to assign or remove services on an existing node in a cluster, and rebalance the cluster.]
:page-topic-type: reference

[abstract]
{description}

[#http-method-and-uri]

=== HTTP Method and URI

----
POST /controller/rebalance
----

== Description

> Use the following REST API to add or remove non-data services on the existing nodes in a cluster.

Reviewer: non-Data instead of non-data and additional descriptions 

Use the following REST API to add or remove non-Data services on the existing nodes in a cluster.   For elements to consider prior to running this command, see the xref:manage:manage-nodes/modify-services-and-rebalance.adoc#prerequisites[Prerequisites] in xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

This REST API reference explains the use of the POST /controller/rebalance command to reconfigure (add or remove) the services on the existing nodes of the cluster.  For the REST API reference for the use of the POST /controller/rebalance command to redistribute services data and processing after failovers or node additions and removals, see xref:rest-api/rest-cluster-rebalance.adoc[Rebalancing the Cluster].

Author: I did not add the following because it is a repetition of the first sentence in this section:
This REST API reference explains the use of the `POST /controller/rebalance` command to reconfigure (add or remove) the services on the existing nodes of the cluster.

Then trigger a rebalance operation.

=== curl Syntax

> You must specify all known nodes and the nodes to be ejected.

Reviewer: You must specify all known nodes in the cluster and the cluster services topology desired.

----
curl -v -X POST -u [admin]:[password] \
http://[localhost]:8091/controller/rebalance \

> [-d knownNodes | -d ejectedNodes | -d topology]

Reviewer: [-d knownNodes | -d topology]

Since the -d ejectedNodes is not needed for this operation -- that option is used in "Rebalancing the Cluster".

----

These are services related parameters:

> * `knownNodes`: The value is a list containing all nodes that are in the cluster at the start of the rebalance process.

Reviewer: `* `knownNodes`: The value is a list containing all nodes that are in the cluster.`

> * `ejectedNodes`: The value is a list containing all nodes that are to be ejected from the cluster during the rebalance process.

Reviewer: Remove line 36.   People can learn about ejectedNodes (if they wish) in the "Rebalancing the Cluster" page.

* `topology`: The value is a list of services and the nodes in which services must reside, after the rebalance operation completes.
+
The value is `topology[<service>]=<list of nodes>`.
+
Where,

** `<service>` can be one of the following services:

*** `fts` for the Search Service.
*** `index` for the Index Service.
*** `n1ql` for the Query Service.
*** `backup` for the Backup Service.
*** `cbas` for the Analytics Service.

** `<list of nodes>` indicate the nodes on which the `<service>` must reside in after the rebalance operation is complete.
The node names are separated by commas.
+
If a `<service>` is running on any node in the cluster that's not in this list, Couchbase Server removes that `<service>` during the rebalance.

> See xref:rest-api/rest-cluster-rebalance.adoc[Rebalancing the Cluster].

Reviewer: I think that this (line 56) is confusing here -- should move it to the bottom where the other references are.

Author: Users need content reference, so I'm leaving it as-is.

** For example, `-d topology[<service>]=otp_node1,otp_node2`.

NOTE: The rebalance occurs even if no `<service>` is changed.

> Each assigned service is allocated a default amount of memory.

Reviewer: Each assigned service is allocated the memory quota for the service currently set on the cluster.

Author: Rephrased it for clarity.

> You can modify these allocations as needed.

Reviewer: You can modify the services memory quotas, if needed, prior to running the services reconfiguration command for the cluster nodes.

| `200 OK` and JSON array containing the expired backups and their details.
| Successful call.

| `400`
| Invalid parameter.

| `400 Object Not found`
| The repository in the endpoint URI does not exist.

| `401 Unauthorized`
|  Authorization failure due to incorrect username or password.

| `403 Forbidden`, plus a JSON message explaining the minimum permissions.
| The provided username has insufficient privileges to call this method.

| `404 Object Not Found`
| Error in the URI path.

| `500 Internal Server Error`
| Error in Couchbase Server.
|===

=== Example

For illustration, consider 5 nodes in the cluster with the following services:

* 172.23.108.69 – Data, Index, Query
* 172.23.108.70 – Data, Search, Query
* 172.23.108.71 – Query
* 172.23.108.72 – Backup
* 172.23.108.73 – Analytics

Reconfigure the services on the existing nodes so that the following is the required output of services on each of the 5 nodes:

* 172.23.108.69 – Data, Query
* 172.23.108.70 – Data, Index
* 172.23.108.71 – Index, Query
* 172.23.108.72 – Search, Query
* 172.23.108.73 – Analytics

Run the following command to reconfigure the services on the existing nodes and get the required output.

> 

Reviewer: Note that since the Backup service is not in the topology list, the Backup service will be removed from the cluster.  Also, there are 2 nodes that did not previously have any services using the disk storage path for Indexes that will have services using the Indexes disk storage path after the cluster topology change has completed with the rebalance.

Author: Rewrote it concisely for clarity.

----
curl -u Administrator:password -X POST 172.23.108.69:8091/controller/rebalance \
-d "knownNodes=ns_1@172.23.108.69,ns_1@172.23.108.70,ns_1@172.23.108.71,ns_1@172.23.108.72,ns_1@172.23.108.73"  \
-d "topology[index]=ns_1@172.23.108.70,ns_1@172.23.108.71"  \
-d "topology[fts]=ns_1@172.23.108.72"  \
-d "topology[n1ql]=ns_1@172.23.108.69,ns_1@172.23.108.71,ns_1@172.23.108.72" \
-d "topology[cbas]=ns_1@172.23.108.73"
---- 

== See Also

* To retrieve the list of services running on a node, see xref:rest-api:rest-list-node-services.adoc[Listing Node Services].

* For information about allocating memory per service, see xref:rest-api:rest-configure-memory.adoc[Configuring Memory].

* To fetch the details of current memory allocation for each service, see xref:rest-api:rest-get-memory-information.adoc[Getting Memory Information].

* For information about assigning services to existing nodes from the UI and CLI, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

> * For information about assigning services to a new node, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Node].

Reviewer: `* For information about assigning services to a new node, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Single Node].`

`* For information about rebalancing, see xref:rest-api/rest-cluster-rebalance.adoc[Rebalancing the Cluster].`

====

=== curl Syntax

You must specify all known nodes in the cluster and the necessary cluster services topology.

----
curl -v -X POST -u [admin]:[password] \
http://[localhost]:8091/controller/rebalance \
[-d knownNodes | -d topology]
----

These are services related parameters:

* `knownNodes`: The value is a list containing all nodes that are in the cluster at the start of the rebalance process.

* `topology`: The value is a list of services and the nodes in which services must reside, after the rebalance operation completes.
+
The value is `topology[<service>]=<list of nodes>`.
+
Where,

** `<service>` can be one of the following services:

*** `fts` for the Search Service.
*** `index` for the Index Service.
*** `n1ql` for the Query Service.
*** `backup` for the Backup Service.
*** `cbas` for the Analytics Service.

** `<list of nodes>` indicate the nodes on which the `<service>` must reside in after the rebalance operation is complete.
The node names are separated by commas.
+
If a `<service>` is running on any node in the cluster that's not in this list, Couchbase Server removes that `<service>` during the rebalance operation.

> For more information, see xref:rest-api:rest-cluster-rebalance.adoc[Rebalancing the Cluster].

Reviewer: Sep 22, 2025 Update:
Remove lines 60 and 61.

** For example, `-d topology[<service>]=otp_node1,otp_node2`.

> 

Reviewer: Sep 22, 2025 Update -- add this additional info:
** To remove (or exclude) one of the above listed services from the cluster, do not list any nodes for that service. For example, `-d topology[<service>]=""`.

== Required Permissions

`Full Admin` role.

[#responses]
== Responses

|===
|Value | Description  

| `200 OK` and JSON array containing the expired backups and their details.
| Successful call.

| `400`
| Invalid parameter.

| `400 Object Not found`
| The repository in the endpoint URI does not exist.

| `401 Unauthorized`
|  Authorization failure due to incorrect username or password.

| `403 Forbidden`, plus a JSON message explaining the minimum permissions.
| The provided username has insufficient privileges to call this method.

| `404 Object Not Found`
| Error in the URI path.

| `500 Internal Server Error`
| Error in Couchbase Server.
|===

=== Example

For illustration, consider 5 nodes in the cluster with the following services:

* 172.23.108.69 – Data, Index, Query
* 172.23.108.70 – Data, Search, Query
* 172.23.108.71 – Query
* 172.23.108.72 – Backup
* 172.23.108.73 – Analytics

Reconfigure the services on the existing nodes so that the following is the required output of services on each of the 5 nodes:

* 172.23.108.69 – Data, Query
* 172.23.108.70 – Data, Index
* 172.23.108.71 – Index, Query
* 172.23.108.72 – Search, Query

> * 172.23.108.73 – Analytics

Reviewer 2: ```suggestion
For example, a cluster with 5 nodes could have the following service configuration:

* 172.23.108.69 – Data, Index, Query
* 172.23.108.70 – Data, Search, Query
* 172.23.108.71 – Query
* 172.23.108.72 – Backup
* 172.23.108.73 – Analytics

You could reconfigure the services on the existing nodes to match the following:

* 172.23.108.69 – Data, Query
* 172.23.108.70 – Data, Index
* 172.23.108.71 – Index, Query
* 172.23.108.72 – Search, Query
* 172.23.108.73 – Analytics
```

> Because the Backup service is not in the topology list, it will be removed from the cluster.

Reviewer: Sep 22, 2025 update:
Remove line 124.

After the topology change and rebalance, two nodes (172.23.108.71 and 172.23.108.72) that previously had no services using the index disk storage path will now have services using it.

> 

Reviewer: Sep 22, 2025 update -- add this additional info:
Also, in the command shown below, the Backup Service is being removed from the cluster, as the `-d topology[backup]=` does not have any nodes listed.

> Run the following command to reconfigure the services on the existing nodes and get the required output.

Reviewer 2: ```suggestion
To reconfigure the 5 nodes to the new topology, run the following command:
```

----
curl -u Administrator:password -X POST 172.23.108.69:8091/controller/rebalance \
-d "knownNodes=ns_1@172.23.108.69,ns_1@172.23.108.70,ns_1@172.23.108.71,ns_1@172.23.108.72,ns_1@172.23.108.73"  \
-d "topology[index]=ns_1@172.23.108.70,ns_1@172.23.108.71"  \
-d "topology[fts]=ns_1@172.23.108.72"  \
-d "topology[n1ql]=ns_1@172.23.108.69,ns_1@172.23.108.71,ns_1@172.23.108.72" \

> -d "topology[cbas]=ns_1@172.23.108.73"

Reviewer: Sep 22, 2025 update -- replace line 135 with the below 2 lines:
-d "topology[backup]=" \
-d "topology[cbas]=ns_1@172.23.108.73"

---- 

== See Also

* To retrieve the list of services running on a node, see xref:rest-api:rest-list-node-services.adoc[Listing Node Services].

* For information about allocating memory per service, see xref:rest-api:rest-configure-memory.adoc[Configuring Memory].

* To fetch the details of current memory allocation for each service, see xref:rest-api:rest-get-memory-information.adoc[Getting Memory Information].

* For information about assigning services to existing nodes from the UI and CLI, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

* For information about assigning services to a new node, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Single Node].

> * For information about rebalancing, see xref:rest-api:rest-cluster-rebalance.adoc[Rebalancing the Cluster].

Reviewer 2: ```suggestion
* For more information about allocating memory per service, see xref:rest-api:rest-configure-memory.adoc[Configuring Memory].

* To fetch the details of current memory allocation for each service, see xref:rest-api:rest-get-memory-information.adoc[Getting Memory Information].

* For more information about assigning services to existing nodes from the UI and CLI, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

* For more information about assigning services to a new node, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Single Node].

* For more information about rebalancing, see xref:rest-api:rest-cluster-rebalance.adoc[Rebalancing the Cluster].
```

= Assigning Services on an Existing Node

:description: pass:q[Use the REST API to assign or remove services on an existing node in a cluster, and rebalance the cluster.]
:page-topic-type: reference

[abstract]
{description}

[#http-method-and-uri]

=== HTTP Method and URI

----
POST /controller/rebalance
----

== Description

Use the following REST API to add or remove non-Data services on the existing nodes in a cluster.

> 

Reviewer: Line 28 needs to be here.  Currently, these two sentences are separated by a NOTE.  It's confusing.  The two sentences go together.

Use the following REST API to add or remove non-Data services on the existing nodes in a cluster.  
Then trigger a rebalance operation.

[NOTE]
====
* Follow the xref:manage:manage-nodes/modify-services-and-rebalance.adoc#prerequisites[Prerequisites] in xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance] before using the REST API.

> * For information about using the `POST /controller/rebalance` REST API to redistribute services data and processing after failovers or node additions and removal, see xref:rest-api:rest-cluster-rebalance.adoc[Rebalancing the Cluster].

Reviewer: Realized that this sentence needed to be more clear.

* For information about using the `POST /controller/rebalance` REST API to rebalance after node additions and removals, after graceful failover & recovery, and after a hard failover, see xref:rest-api:rest-cluster-rebalance.adoc[Rebalancing the Cluster].

====

> Then trigger a rebalance operation.

Reviewer: Move this line to be after line 19.

=== curl Syntax

> You must specify all known nodes in the cluster and the necessary cluster services topology.

Reviewer: You must specify all known nodes in the cluster and the desired cluster services topology.

|Value | Description  

| `200 OK` and JSON array containing the expired backups and their details.
| Successful call.

| `400`
| Invalid parameter.

| `400 Object Not found`
| The repository in the endpoint URI does not exist.

| `401 Unauthorized`
|  Authorization failure due to incorrect username or password.

| `403 Forbidden`, plus a JSON message explaining the minimum permissions.
| The provided username has insufficient privileges to call this method.

| `404 Object Not Found`
| Error in the URI path.

| `500 Internal Server Error`
| Error in Couchbase Server.
|===

=== Example

For illustration, consider 5 nodes in the cluster with the following services:

* 172.23.108.69 – Data, Index, Query
* 172.23.108.70 – Data, Search, Query
* 172.23.108.71 – Query
* 172.23.108.72 – Backup
* 172.23.108.73 – Analytics

Reconfigure the services on the existing nodes so that the following is the required output of services on each of the 5 nodes:

* 172.23.108.69 – Data, Query
* 172.23.108.70 – Data, Index
* 172.23.108.71 – Index, Query
* 172.23.108.72 – Search, Query
* 172.23.108.73 – Analytics

Because the Backup service is not in the topology list, it will be removed from the cluster.

> After the topology change and rebalance, two nodes that previously had no services using the index disk storage path will now have services using it.

Reviewer: After the topology change and rebalance, two nodes (172.23.108.71 and 72), that previously had no services using the index disk storage path, will now have services using it.
